From adcc9d700b413da60af7ad113e31d2b919e32482 Mon Sep 17 00:00:00 2001
From: wu xia <xia.wu@intel.com>
Date: Tue, 26 Apr 2016 15:31:41 +0300
Subject: [PATCH 1402/2302] media: Only requests in IDLE state may be deleted

Prevent deleting queued requests. Also mark deleted requests as such
by adding a new state for them.

Signed-off-by: Sakari Ailus <sakari.ailus@linux.intel.com>
Signed-off-by: wu xia <xia.wu@intel.com>
---
 drivers/media/media-device.c | 21 +++++++++++++++++----
 include/media/media-device.h |  1 +
 2 files changed, 18 insertions(+), 4 deletions(-)

diff --git a/drivers/media/media-device.c b/drivers/media/media-device.c
index e82989b..9154ea6 100644
--- a/drivers/media/media-device.c
+++ b/drivers/media/media-device.c
@@ -38,6 +38,7 @@
 static char *__request_state[] = {
 	"IDLE",
 	"QUEUED",
+	"DELETED",
 };
 
 #define request_state(i)			\
@@ -166,12 +167,22 @@ static int media_device_request_alloc(struct media_device *mdev,
 	return ret;
 }
 
-static void media_device_request_delete(struct media_device *mdev,
-					struct media_device_request *req)
+static int media_device_request_delete(struct media_device *mdev,
+				       struct media_device_request *req)
 {
 	unsigned long flags;
 
 	spin_lock_irqsave(&mdev->req_lock, flags);
+
+	if (req->state != MEDIA_DEVICE_REQUEST_STATE_IDLE) {
+		spin_unlock_irqrestore(&mdev->req_lock, flags);
+		dev_dbg(mdev->dev, "request: can't delete %u, state %s\n",
+			req->id, request_state(req->state));
+		return -EINVAL;
+	}
+
+	req->state = MEDIA_DEVICE_REQUEST_STATE_DELETED;
+
 	if (req->filp) {
 		/*
 		 * If the file handle is gone by now the
@@ -182,9 +193,12 @@ static void media_device_request_delete(struct media_device *mdev,
 		list_del(&req->fh_list);
 		req->filp = NULL;
 	}
+
 	spin_unlock_irqrestore(&mdev->req_lock, flags);
 
 	media_device_request_put(req);
+
+	return 0;
 }
 
 static int media_device_request_queue_apply(
@@ -250,8 +264,7 @@ static long media_device_request_cmd(struct media_device *mdev,
 		break;
 
 	case MEDIA_REQ_CMD_DELETE:
-		media_device_request_delete(mdev, req);
-		ret = 0;
+		ret = media_device_request_delete(mdev, req);
 		break;
 
 	case MEDIA_REQ_CMD_APPLY:
diff --git a/include/media/media-device.h b/include/media/media-device.h
index bba7dcf..4e9b1db 100644
--- a/include/media/media-device.h
+++ b/include/media/media-device.h
@@ -34,6 +34,7 @@ struct media_device;
 enum media_device_request_state {
 	MEDIA_DEVICE_REQUEST_STATE_IDLE,
 	MEDIA_DEVICE_REQUEST_STATE_QUEUED,
+	MEDIA_DEVICE_REQUEST_STATE_DELETED,
 };
 
 /**
-- 
2.7.4

